memory_unsafe.go

```diff
@@ -0,0 +1,256 @@
+// through another object (like a sync.Atomic) that can potentially be passed
+// around the Go application. Accidentally losing the reference to the *Memory
+// would result in hard-to-debug segfaults, which are always unexpected in Go.
+
+//go:linkname heapObjectsCanMove runtime.heapObjectsCanMove
+func heapObjectsCanMove() bool
+
+// ErrInvalidType is returned when the given type cannot be used as a Memory or
+// Variable pointer.
+var ErrInvalidType = errors.New("invalid type")
+
+func newUnsafeMemory(fd, size int) (*Memory, error) {
+	// Some architectures need the size to be page-aligned to work with MAP_FIXED.
+	if internal.Align(size, os.Getpagesize()) != size {
+		return nil, fmt.Errorf("memory: must be a multiple of page size (requested %d bytes)", size)
+	}
+
+	// Allocate a page-aligned span of memory on the Go heap.
+	alloc, err := allocate(size)
+	if err != nil {
+		return nil, fmt.Errorf("allocating memory: %w", err)
+	}
+
+	// Typically, maps created with BPF_F_RDONLY_PROG remain writable from user
+	// space until frozen. As a security precaution, the kernel doesn't allow
+	// mapping bpf map memory as read-write into user space if the bpf map was
+	// frozen, or if it was created using the RDONLY_PROG flag.
+	//
+	// The user would be able to write to the map after freezing (since the kernel
+	// can't change the protection mode of an already-mapped page), while the
+	// verifier assumes the contents to be immutable.
+	//
+	// Map the bpf map memory over a page-aligned allocation on the Go heap.
+	err = mapmap(fd, alloc, size, unix.PROT_READ|unix.PROT_WRITE)
+
+	// If the map is frozen when an rw mapping is requested, expect EPERM. If the
+	// map was created with BPF_F_RDONLY_PROG, expect EACCES.
+	var ro bool
+	if errors.Is(err, unix.EPERM) || errors.Is(err, unix.EACCES) {
+		ro = true
+		err = mapmap(fd, alloc, size, unix.PROT_READ)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("setting up memory-mapped region: %w", err)
+	}
+
+	mm := &Memory{
+		unsafe.Slice((*byte)(alloc), size),
+		ro,
+		true,
+	}
+
+	return mm, nil
+}
+
+// allocate returns a pointer to a page-aligned section of memory on the Go
+// heap, managed by the runtime.
+func allocate(size int) (unsafe.Pointer, error) {
+	// Memory-mapping over a piece of the Go heap is unsafe when the GC can
+	// randomly decide to move objects around, in which case the mapped region
+	// will not move along with it.
+	if heapObjectsCanMove() {
+		return nil, errors.New("this Go runtime has a moving garbage collector")
+	}
+
+	if size == 0 {
+		return nil, errors.New("size must be greater than 0")
+	}
+
+	// Request at least two pages of memory from the runtime to ensure we can
+	// align the requested allocation to a page boundary. This is needed for
+	// MAP_FIXED and makes sure we don't mmap over some other allocation on the Go
+	// heap.
+	size = internal.Align(size+os.Getpagesize(), os.Getpagesize())
```

Comment:
Something we discussed on the previous PR I think: this will be expensive on platforms that have large page sizes like arm64. Not sure there is anything we can do about it, but might make sense to call it out in the docs.